File: ios/Classes/CoordinateMapper.swift
import CoreGraphics
import ImageIO

/// Converts Vision's normalized, bottom-left-origin geometry into the
/// pixel-space, top-left-origin representation the Dart layer expects.
///
/// Frames are handed to Vision already oriented upright (via
/// `CGImagePropertyOrientation`), so the reported `imageSize` is the oriented
/// size and `rotation` is always 0 on iOS — the Flutter overlay then maps
/// detections to the upright preview with no extra rotation.
enum CoordinateMapper {

  /// Oriented (upright) pixel size for a buffer of [width]x[height].
  static func orientedSize(
    width: Int, height: Int, orientation: CGImagePropertyOrientation
  ) -> CGSize {
    switch orientation {
    case .left, .right, .leftMirrored, .rightMirrored:
      return CGSize(width: height, height: width)
    default:
      return CGSize(width: width, height: height)
    }
  }

  /// True when the center of a Vision-normalized [box] is within [area].
  static func isInside(box: CGRect, area: ScanAreaConfig) -> Bool {
    if area.isFull { return true }
    let nx = box.midX
    // Flip Y to top-left origin.
    let ny = 1.0 - box.midY
    return nx >= area.left && nx <= area.left + area.width
      && ny >= area.top && ny <= area.top + area.height
  }

  /// Converts a Vision-normalized point (bottom-left origin) to a pixel point
  /// (top-left origin) within an image of [size].
  static func toPixel(_ p: CGPoint, size: CGSize) -> [String: Double] {
    [
      "x": Double(p.x) * Double(size.width),
      "y": Double(1.0 - p.y) * Double(size.height),
    ]
  }

  /// Converts a Vision-normalized rect to a pixel rect (top-left origin).
  static func toPixelRect(_ r: CGRect, size: CGSize) -> [String: Double] {
    let left = Double(r.minX) * Double(size.width)
    let top = Double(1.0 - r.maxY) * Double(size.height)
    return [
      "left": left,
      "top": top,
      "width": Double(r.width) * Double(size.width),
      "height": Double(r.height) * Double(size.height),
    ]
  }
}

File: android/src/main/kotlin/com/karnival/barcode_scanner_pro/CoordinateMapper.kt
package com.karnival.barcode_scanner_pro

import android.graphics.Rect

/**
 * Maps between ML Kit's unrotated image-buffer coordinate space and the upright,
 * normalized preview space that the Dart scan area is expressed in.
 *
 * The scan area arrives from Dart as a `[0,1]` rectangle relative to the upright
 * preview. ML Kit reports detections in the raw buffer (pre-rotation) space, so
 * we rotate a detection's center into upright space, normalize it by the
 * orientation-corrected image size, and test containment. This avoids any
 * per-frame bitmap cropping/allocation while still honoring the scan area.
 */
internal object CoordinateMapper {

    /** True when the center of [box] lies within the normalized [area]. */
    fun isInside(
        box: Rect,
        area: ScanAreaConfig,
        imageWidth: Int,
        imageHeight: Int,
        rotation: Int,
    ): Boolean {
        if (area.isFull) return true

        val cx = box.exactCenterX()
        val cy = box.exactCenterY()

        // Rotate the point into upright space and compute upright dimensions.
        val (ux, uy, uw, uh) = rotatePoint(cx, cy, imageWidth, imageHeight, rotation)

        val nx = ux / uw
        val ny = uy / uh
        return nx >= area.left && nx <= area.left + area.width &&
            ny >= area.top && ny <= area.top + area.height
    }

    private data class Mapped(val x: Float, val y: Float, val w: Float, val h: Float)

    private operator fun Mapped.component1() = x
    private operator fun Mapped.component2() = y
    private operator fun Mapped.component3() = w
    private operator fun Mapped.component4() = h

    private fun rotatePoint(
        x: Float,
        y: Float,
        w: Int,
        h: Int,
        rotation: Int,
    ): Mapped = when (rotation) {
        90 -> Mapped(h - y, x, h.toFloat(), w.toFloat())
        180 -> Mapped(w - x, h - y, w.toFloat(), h.toFloat())
        270 -> Mapped(y, w - x, h.toFloat(), w.toFloat())
        else -> Mapped(x, y, w.toFloat(), h.toFloat())
    }
}
